tools/multi_loader/runner/multi_loader_runner.go

```
@@ -0,0 +1,411 @@
+
+	// Create output directory
+	outputDir := path.Dir(experimentConfig.OutputPathPrefix)
+
+	if err := os.MkdirAll(outputDir, 0755); err != nil {
+		log.Fatal(err)
+	}
+	// Write experiment configs to temp file
+	d.writeExperimentConfigToTempFile(experimentConfig, EXPERIMENT_TEMP_CONFIG_PATH)
+}
+
+/**
+* Merge base configs with partial loader configs
+ */
+func (d *MultiLoaderRunner) mergeConfigurations(baseConfigPath string, experiment types.LoaderExperiment) config.LoaderConfiguration {
+	// Read base configuration
+	baseConfigByteValue, err := os.ReadFile(baseConfigPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Debug("Experiment configuration ", experiment.Config)
+
+	var mergedConfig config.LoaderConfiguration
+	// Unmarshal base configuration
+	if err = json.Unmarshal(baseConfigByteValue, &mergedConfig); err != nil {
+		log.Fatal(err)
+	}
+
+	log.Debug("Base configuration ", mergedConfig)
+
+	// merge experiment config onto base config
+	experimentConfigBytes, _ := json.Marshal(experiment.Config)
+	if err = json.Unmarshal(experimentConfigBytes, &mergedConfig); err != nil {
+		log.Fatal(err)
+	}
+	log.Debug("Merged configuration ", mergedConfig)
+
+	return mergedConfig
+}
+
+func (d *MultiLoaderRunner) writeExperimentConfigToTempFile(experimentConfig config.LoaderConfiguration, fileWritePath string) {
+	experimentConfigBytes, _ := json.Marshal(experimentConfig)
+	err := os.WriteFile(fileWritePath, experimentConfigBytes, 0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (d *MultiLoaderRunner) runExperiment(experiment types.LoaderExperiment) error {
+	log.Debug("Experiment configuration ", experiment.Config)
+
+	// Create the log file
+	logFilePath := path.Join(path.Dir(experiment.Config["OutputPathPrefix"].(string)), "loader.log")
+	logFile, err := os.OpenFile(logFilePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer logFile.Close()
+
+	for i := 0; i < NUM_OF_RETRIES; i++ {
```

You can simplify the body a bit: compute the actual number of retries based on fail fast and dry run before the loop. In the body, on every iteration except the first, you log that it is retry; after that, you run execute the loader command and check for success, if so, return as completed. If we exit the loop, it means we failed, so move the handling code outside.

---

That's a good point! Will make the update. Also currently, NUM_OF_RETIRES is used as the total number of executions which is confusing will update the logic for that too.